crates/wasm-smith/src/core.rs

```diff
@@ -1073,7 +1109,16 @@ impl Module {
 
         if self.config.gc_enabled && concrete_type_limit > 0 && u.arbitrary()? {
             let idx = u.int_in_range(0..=concrete_type_limit - 1)?;
-            return Ok(HeapType::Concrete(idx));
+            // If the caller is demanding a shared heap type but the concrete
+            // type we found is not in fact shared, we skip down below to use an
+            // abstract heap type instead. If the caller is not demanding a
+            // shared type, though, we can use either a shared or unshared
+            // concrete type.
+            if let Some(ty) = self.types.get(idx as usize) {
+                if !(self.must_share && !ty.composite_type.shared) {
+                    return Ok(HeapType::Concrete(idx));
+                }
+            }
```

Comment:
This might be good to haev as a TODO where in the future this consults the list of shared types instead of "all types" to avoid this arm entirely if there are no shared types.